### ISSUE TITLE
Field validator should check for ECS fields marked as arrays

### ISSUE THREAD

**Issue by andrewkroh** [elastic/elastic-package#615]
It would be helpful if elastic-package could trigger errors on ECS event fields that are not arrays. For example, [host.ip](https://www.elastic.co/guide/en/ecs/current/ecs-host.html#field-host-ip) is specified to be an array so events that contain the field should be an array of ips.

<img width="745" alt="Screen Shot 2021-12-08 at 8 49 04 AM" src="https://user-images.githubusercontent.com/4565752/145219412-387a58ea-a250-494e-9d15-de5d35623809.png">

The generated YAML specification indicates the array type. https://github.com/elastic/ecs/blob/9b463c9116eb594a3be10db01c3236fd6b44b996/generated/ecs/ecs_flat.yml#L4536-L4543

```yaml
host.ip:
  dashed_name: host-ip
  description: Host ip addresses.
  flat_name: host.ip
  level: core
  name: ip
  normalize:
  - array
  short: Host ip addresses.
  type: ip
```

**Comment by andrewkroh** [comment]
Another validation that would be nice to apply is for the `allowed_values` listed in the schema for fields like `event.{kind,type,category}`.

https://github.com/elastic/ecs/blob/4cbe6f7f9cb78160153659802eb66dd135fb4677/schemas/event.yml#L75-L76

**Comment by andrewkroh** [comment]
`pattern` has now been added to the ECS schema. See https://github.com/elastic/ecs/issues/1741 https://github.com/elastic/ecs/pull/1834.

So the `elastic-package` tool could now check data against the `pattern` defined in ECS for a given field.

**Comment by jsoriano** [comment]
Hey @andrewkroh,

I am implementing the check for arrays in https://github.com/elastic/elastic-package/pull/765, and I have found that pipeline tests of several packages in the integrations repository (35 out of 140) would fail if we enable this. So I wonder if we should do it.

What would be the advantage of checking that fields are formatted as arrays? In principle for Elasticsearch it is the same.

**Comment by andrewkroh** [comment]
> What would be the advantage of checking that fields are formatted as arrays? In principle for Elasticsearch it is the same.

You're right, for Elasticsearch indexing it doesn't make a difference whether the field is a scalar or an array. ECS added an array indicator to the schema primarily for things other than Elasticsearch that might be producing or consuming the data.

I think it would be advantageous if the `_source` data in events that claim to use ECS fields actually matches the field definition. That consistency makes it easier for anything that's trying to consume the ECS data. That includes the developers of ingest pipelines as they move between various integrations. And it's one less thing I need to spend time manually inspecting during integration PR reviews.

One of the earliest issues I recall while implementing ECS and having consumers in Kibana was with TypeScipt code in Kibana that was generated/derived from ECS. There was a mismatch between the _source data and the generated types due to arrays. That code now handles both IIRC, but it would nice if we were consistent with our types so future developers can write simpler code to read/process ECS data.

> (35 out of 140) would fail if we enable this

Is there a way we can enable this in permissive manner while we fix those 35? It's kind of a general problem for any case where we enable stricter validation (like checking that `event.*` is defined in fields).

**Comment by jsoriano** [comment]
> That consistency makes it easier for anything that's trying to consume the ECS data.

I was wondering if Elasticsearch provides something to help here, and I have found that there is a meta issue about improving the situation for single vs multi-valued fields. https://github.com/elastic/elasticsearch/issues/80825

> Is there a way we can enable this in permissive manner while we fix those 35? It's kind of a general problem for any case where we enable stricter validation (like checking that `event.*` is defined in fields).

There isn't at the moment, but this is something that has already appeared in other discussions. Marcin recently open an issue about different validation modes (https://github.com/elastic/package-spec/issues/313).

I will put this issue on hold till we have something like this.

**Comment by jsoriano** [comment]
Implemented as part of package spec 2.0.0. Packages will be able to fix normalization issues as they upgrade to this format.
